idx/memory/write_queue.go

```diff
@@ -0,0 +1,106 @@
+package memory
+
+import (
+	"sync"
+	"time"
+
+	"github.com/grafana/metrictank/idx"
+	"github.com/raintank/schema"
+)
+
+type WriteQueue struct {
+	sync.RWMutex
+	archives    map[schema.MKey]*idx.Archive
+	shutdown    chan struct{}
+	done        chan struct{}
+	maxBuffered int
+	maxDelay    time.Duration
+	flushed     chan struct{}
+
+	idx *UnpartitionedMemoryIdx
+}
+
+// NewWriteQueue creates a new writeQueue that will add archives to the passed UnpartitionedMemoryIdx
+// in batches
+func NewWriteQueue(index *UnpartitionedMemoryIdx, maxDelay time.Duration, maxBuffered int) *WriteQueue {
+	wq := &WriteQueue{
+		archives:    make(map[schema.MKey]*idx.Archive),
+		shutdown:    make(chan struct{}),
+		done:        make(chan struct{}),
+		maxBuffered: maxBuffered,
+		maxDelay:    maxDelay,
+		flushed:     make(chan struct{}, 1),
+		idx:         index,
+	}
+	go wq.loop()
+	return wq
+}
+
+func (wq *WriteQueue) Stop() {
+	close(wq.shutdown)
+	<-wq.done
+}
+
+func (wq *WriteQueue) Queue(archive *idx.Archive) {
+	wq.Lock()
+	wq.archives[archive.Id] = archive
+	if len(wq.archives) >= wq.maxBuffered {
+		wq.flush()
+	}
+	wq.Unlock()
+}
+
+func (wq *WriteQueue) Get(id schema.MKey) (*idx.Archive, bool) {
+	wq.RLock()
+	a, ok := wq.archives[id]
+	wq.RUnlock()
+	return a, ok
+}
+
+// flush adds the buffered archives to the memoryIdx.
+// callers need to acquire a writeLock before calling this function.
+func (wq *WriteQueue) flush() {
+	if len(wq.archives) == 0 {
+		// non blocking write to the flushed chan.
+		// if we cant write to the flushed chan it means there is a previous flush
+		// signal that hasnt been processed.  In that case, we dont need to send another one.
+		select {
+		case wq.flushed <- struct{}{}:
+		default:
+		}
+		return
+	}
+	wq.idx.Lock()
+	for _, archive := range wq.archives {
+		wq.idx.add(archive)
```

Comment:
so in `.add()` we're taking this pointer and assigning it to `m.defById` without every copying the values. this means we keep a reference to the underlying byte array of this buffer slice. if in the future a lot of those `*idx.Archive` that have been added via the buffer get pruned again because they become inactive, then the underlying buffer slice will only get freed once all that have been in the buffer have been pruned. So i think it might be worth it to copy that value once instead of passing a direct reference to the slice element or not?